### PR TITLE
Set instance state to failed on bad instance_put

### PR DESCRIPTION
**Pull request by jmpesp**
If a call to sled-agent's `instance_put` fails with anything other than a 400 (bad request), set instance's state to failed - we cannot know what state the instance is in.

This exposed another two problems:

- Nexus would only delete network interfaces off instances that were stopped
- Nexus would only detach disks off instances that were creating or stopped.

This commit changes the relevant network interface queries to allow deletion of network interfaces for instances that are either stopped or failed, and adds Failed to the list of ok_to_detach_instance_states for disks. Note that network interface insertion still requires an instance to be stopped, not failed.

Closes https://github.com/oxidecomputer/omicron/issues/1713

### CODE REVIEW

File: sled-agent/src/sim/sled_agent.rs [hunk 1]
```diff
@@ -194,6 +194,13 @@ impl SledAgent {
         initial_hardware: InstanceHardware,
         target: InstanceRuntimeStateRequested,
     ) -> Result<InstanceRuntimeState, Error> {
+        let ncpus: i64 = (&initial_hardware.runtime.ncpus).into();
+        if ncpus > 16 {
+            return Err(Error::internal_error(
```
Comment (jmpesp) [inline sled-agent/src/sim/sled_agent.rs:199]: I know this should probably be `bad_request` instead, but for `test_instance_fails_to_boot_with_disk` I needed it to be an `internal_error`. I welcome any input for better ways to do this  :)

File: nexus/src/app/instance.rs [hunk 2]
```diff
@@ -604,25 +604,84 @@ impl super::Nexus {
...
+                // The sled-agent has told us that it can't do what we
+                // requested, but does that mean a failure? One example would be
+                // if we try to "reboot" a stopped instance. That shouldn't
+                // transition the instance to failed. But if the sled-agent
+                // *can't* boot a stopped instance, that should transition
+                // to failed.
+                //
+                // Without a richer error type, let the sled-agent tell Nexus
+                // what to do with status codes.
+                error!(self.log, "saw {} from instance_put!", &e,);
```
Comment (luqmana) [inline nexus/src/app/instance.rs:642]:
```suggestion
                error!(self.log, "saw {} from instance_put!", e);
```

File: nexus/src/app/instance.rs [hunk 3]
```diff
@@ -604,25 +604,84 @@ impl super::Nexus {
...
+
+                // this is unfortunate, but sled_agent_client::Error doesn't
+                // implement Copy, and can't be match'ed upon below without this
+                // line.
+                let e = e.into();
```
Comment (luqmana) [inline nexus/src/app/instance.rs:647]: Does the Copy bit actually matter? It'd doesn't seem like you need to copy anywhere, and the `.into()` is gonna be needed regardless.

for e.g. you could do:
```rust
match e {
    e @ sled_agent_client::Error::InvalidRequest { .. } => Err(e.into()),
    e => {
        // ...
       Err(e.into())
    }
}
```

or just hoist the `.into()`:
```rust
match e.into() {
    e @ Error::InvalidRequest { .. } => Err(e),
    e => {
        // ...
       Err(e)
    }
}
```

File: nexus/tests/integration_tests/instances.rs [hunk 4]
```diff
@@ -1798,6 +1803,99 @@ async fn test_attach_one_disk_to_instance(cptestctx: &ControlPlaneTestContext) {
...
+
+    let url_instances = format!(
+        "/organizations/{}/projects/{}/instances",
+        ORGANIZATION_NAME, PROJECT_NAME
+    );
+
+    let builder =
+        RequestBuilder::new(client, http::Method::POST, &url_instances)
+            .body(Some(&instance_params))
+            .expect_status(Some(http::StatusCode::INTERNAL_SERVER_ERROR));
```
Comment (luqmana) [inline nexus/tests/integration_tests/instances.rs:1872]: What will actually cause this instance create to fail? The setup looks the same as in `test_attach_one_disk_to_instance` above which should succeed.

Comment (luqmana) [reply nexus/tests/integration_tests/instances.rs:1872]: Ah, I see the `> 16` vCPUs bit :P Add a comment to indicate why this should fail here?

File: sled-agent/src/sim/sled_agent.rs (hunk 1 above)
Comment (luqmana) [reply sled-agent/src/sim/sled_agent.rs:199]: First thought was `InvalidValue` but that also just gets mapped to 400 as well :/

Either way, add a comment here to indicate why this specific error

**Review by luqmana — APPROVED** [review]
Small nits but thanks!

I think I might've inadvertently exposed this by changing exactly when an instance transitions from `Creating` to `Starting`.

File: nexus/src/app/instance.rs (hunk 3 above)
Comment (jmpesp) [reply nexus/src/app/instance.rs:647]: I've tried all variants of moving around the borrow, and `.into()`, and what's there is the only thing I could make work unfortunately.

Comment (luqmana) [reply nexus/src/app/instance.rs:647]: you shouldn't need a borrow at all?

File: sled-agent/src/sim/sled_agent.rs (hunk 1 above)
Comment (jmpesp) [reply sled-agent/src/sim/sled_agent.rs:201]: added comments in be9f7a37

File: nexus/src/app/instance.rs (hunk 2 above)
Comment (jmpesp) [reply nexus/src/app/instance.rs]: removed in 6e85d66e5fc10d911fcbb55f432bf854e6adc6e5

File: nexus/tests/integration_tests/instances.rs (hunk 4 above)
Comment (jmpesp) [reply nexus/tests/integration_tests/instances.rs:1877]: added comments in be9f7a37

**Review by bnaecker — APPROVED** [review]
Looks good, thanks for doing this!

**Pull request merged**